api/models/v0.3/modelCard.js

```diff
@@ -407,17 +418,28 @@ class Card {
    * If the current card is valid, then create it against the ILS API.
    */
   async createIlsPatron() {
-    if (this.policy.isRequiredField(this.barcode)) {
-      this.setBarcode();
-    }
+    let response;
     this.setPtype();
 
     if (!this.validForIls()) {
-      throw new Error("Some error with ILS");
+      throw new Error("The card has not been validated or has no ptype.");
+    }
+
+    // For patrons with the `simplye` policy type, the barcode is required,
+    // so let's create one.
+    if (this.policy.isRequiredField("barcode")) {
+      await this.setBarcode();
     }
 
-    // create the patron
-    const response = await this.ilsClient.createPatron(this);
+    // Now create the patron in the ILS.
+    try {
+      response = await this.ilsClient.createPatron(this);
+    } catch (error) {
+      // We want to catch the error here just to free up the barcode in the
+      // database, but continue to throw the same error as an API response.
+      await this.freeBarcode(this.barcode);
```

Comment:
I'm also concerned about a race condition here. Let's imagine two processes running at once, A and B.

They both get the idea that the next available barcode is 23333567.

Process A creates a patron with that barcode and marks the barcode as used.
Process B tries to create a patron with the same barcode and gets an error -- the patron already exists.
Process B then frees the barcode because it encountered an error.

Now it looks like 23333567 is free for some future process to use, even though there's a patron with that barcode.

Reply:
I updated the database checking and setting of `used` based on the other comments. _I think_ the next commit might help fix this issue.

If the barcode is new, an error will be thrown when it's inserted into the database if it's already there and a new barcode will be used.

If the barcode is an existing unused barcode, then an error will be thrown when updating the value of `unused` to true. So if A uses the barcode it's set to true and if B tries to set it to true and it's already true, then it will try a new barcode.